processing/src/main/java/org/apache/druid/segment/join/filter/rewrite/JoinFilterPreAnalysisGroup.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.join.filter.rewrite;
+
+import org.apache.druid.query.filter.Filter;
+import org.apache.druid.segment.join.filter.JoinFilterPreAnalysis;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A JoinFilterPreAnalysisGroup holds all of the JoinFilterPreAnalysis objects for a given query and
+ * also stores the per-query parameters that control the filter rewrite operations (from the query context).
+ *
+ * The analyses map is keyed by Filter: each Filter in the map belongs to separate level of query
+ * (e.g. outer query, subquery level 1, etc.)
+ *
+ * A concurrent hash map is used since the per-level Filter processing occurs across multiple threads.
+ */
+public class JoinFilterPreAnalysisGroup
+{
+  private final boolean enableFilterPushDown;
+  private final boolean enableFilterRewrite;
+  private final boolean enableRewriteValueColumnFilters;
+  private final long filterRewriteMaxSize;
+  private final ConcurrentHashMap<Filter, JoinFilterPreAnalysis> analyses;
```

Comment:
Have you looked at the implementations of hashCode and equals for all filters? Are they efficient? What happens if the filter is something like an IN filter with a very large list of values? The hashCode check could be slower than deciding that the filter can not be pushed down.

Reply:
The main one I would be concerned with is InFilter, and to a lesser extent AND/OR.

I added lazy computation/caching for InFilter/AndFilter/OrFilter hashcodes, as long as the hashing is done once per query, and not once per segment, I wouldn't expect hashing overhead to outweigh the benefits of the rewrite (esp since such large filters would be expensive to apply per-row on the RHS)